samples/vboxwrapper/vboxwrapper.cpp

```diff
@@ -906,95 +871,37 @@ int main(int argc, char** argv) {
-                "   NOTE: VirtualBox has reported an improperly configured virtual machine. It was configured to require\n"
-                "    hardware acceleration for virtual machines, but your processor does not support the required feature.\n"
-                "    Please report this issue to the project so that it can be addresssed.\n";
+            error_reason = pVM->get_error(NO_HA);
         } else if (pVM->is_logged_failure_vm_extensions_in_use()) {
-            error_reason =
-                "   NOTE: VirtualBox hypervisor reports that another hypervisor has locked the hardware acceleration\n"
-                "    for virtual machines feature in exclusive mode.\n";
+            error_reason = pVM->get_error(LOCKED_HA);
             unrecoverable_error = false;
-            temp_reason = "Forign VM Hypervisor locked hardware acceleration features.";
+            temp_reason = pVM->get_error(FOREIGN_HYPERV);
             temp_delay = 86400;
         } else if (pVM->is_logged_failure_host_out_of_memory()) {
-            error_reason =
-                "   NOTE: VirtualBox has failed to allocate enough memory to start the configured virtual machine.\n"
-                "    This might be a temporary problem and so this job will be rescheduled for another time.\n";
+            error_reason = pVM->get_error(NO_MEM);
             unrecoverable_error = false;
-            temp_reason = "VM Hypervisor was unable to allocate enough memory to start VM.";
+            temp_reason = pVM->get_error(TEMP_NO_MEM);
         } else if (timeout <= dtime()) {
-            error_reason =
-                "   NOTE: VM failed to enter an online state within the timeout period.\n"
-                "    This might be a temporary problem and so this job will be rescheduled for another time.\n";
+            error_reason = pVM->get_error(NOT_ONLINE);
             unrecoverable_error = false;
             do_dump_hypervisor_logs = true;
-            temp_reason = "VM Hypervisor failed to enter an online state in a timely fashion.";
+            temp_reason = pVM->get_error(NO_ONLINE);
             temp_delay = 86400;
         }
 
-        if (unrecoverable_error) {
-            // Attempt to cleanup the VM and exit.
-            if (!skip_cleanup) {
-                pVM->cleanup();
-            }
-
-            checkpoint.update(elapsed_time, current_cpu_time);
-
-            if (error_reason.size()) {
-                vboxlog_msg("\n%s", error_reason.c_str());
-            }
+        pVM->report_clean(unrecoverable_error, skip_cleanup, do_dump_hypervisor_logs,
+                retval, error_reason, pVM->vm_pid, temp_delay, temp_reason,
+                current_cpu_time, last_checkpoint_cpu_time, fraction_done,
+                bytes_sent, bytes_received);
 
-            if (do_dump_hypervisor_logs) {
-                pVM->dump_hypervisor_logs(true);
-            }
+        if (unrecoverable_error) checkpoint.update(elapsed_time, current_cpu_time);
```

Comment:
This code needs to go before `report_clean` because `report_clean` doesn't return.